// @azure/storage-blob/dist-esm/storage-blob/src/utils/utils.browser.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
/**
 * Convert a Browser Blob object into ArrayBuffer.
 *
 * @param blob -
 */
export async function blobToArrayBuffer(blob) {
    const fileReader = new FileReader();
    return new Promise((resolve, reject) => {
        fileReader.onloadend = (ev) => {
            resolve(ev.target.result);
        };
        fileReader.onerror = reject;
        fileReader.readAsArrayBuffer(blob);
    });
}
/**
 * Convert a Browser Blob object into string.
 *
 * @param blob -
 */
export async function blobToString(blob) {
    const fileReader = new FileReader();
    return new Promise((resolve, reject) => {
        fileReader.onloadend = (ev) => {
            resolve(ev.target.result);
        };
        fileReader.onerror = reject;
        fileReader.readAsText(blob);
    });
}
export function streamToBuffer() { }
export function streamToBuffer2() { }
export function readStreamToLocalFile() { }
export const fsStat = function stat() { };
export const fsCreateReadStream = function createReadStream() { };
